sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,346 @@
+
+            return new PooledProducer(item.PartitionProducer, cleanUp: producer =>
+            {
+                Argument.AssertNotNull(producer, nameof(producer));
+
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.ExtendRemoveAfter(removeAfterDuration);
+                }
+
+                // If TryRemove returned false the PoolItem would not be closed deterministically
+                // and the ExpirationTimer callback would eventually remove it from the
+                // Pool leaving to the Garbage Collector the responsability of closing
+                // the TransportProducer and the AMQP link.
+
+                item.ActiveInstances.TryRemove(identifier, out _);
+
+                // The second TryGetValue runs after the extension would have been seen, so it
+                // is intended to be sure that the item wasn't removed in the meantime.
+
+                if (!Pool.TryGetValue(partitionId, out _) && !item.ActiveInstances.Any())
+                {
+                    return producer.CloseAsync(CancellationToken.None);
+                }
+
+                return Task.CompletedTask;
+            });
+        }
+
+        /// <summary>
+        ///   Performs the task needed to clean up resources used by the <see cref="TransportProducerPool" />.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async ValueTask DisposeAsync() => await CloseAsync().ConfigureAwait(false);
+
+        /// <summary>
+        ///   Closes the producers in the pool and performs any cleanup necessary
+        ///   for resources used by the <see cref="TransportProducerPool" />.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task CloseAsync(CancellationToken cancellationToken = default)
+        {
+            await EventHubProducer.CloseAsync(cancellationToken).ConfigureAwait(false);
+
+            var pendingCloses = new List<Task>();
+
+            foreach (var poolItem in Pool.Values)
+            {
+                pendingCloses.Add(poolItem.PartitionProducer.CloseAsync(CancellationToken.None));
+            }
+
+            Pool.Clear();
+            ExpirationTimer.Dispose();
+
+            await Task.WhenAll(pendingCloses).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///   Returns a <see cref="TimerCallback" /> that will search for all the expired <see cref="PoolItem" />
+        ///   in the <see cref="Pool" /> and will try to close those that have expired.
+        /// </summary>
+        ///
+        /// <returns>A <see cref="TimerCallback" /> that is periodically run every <see cref="DefaultPerformExpirationPeriod" />.</returns>
+        ///
+        internal TimerCallback PerformExpiration()
```

Comment:
nit: We should probably call this something like `CreateExpirationTimerCallback` or similar to indicate that it doesn't actually perform the expiration but creates a callback that does.